src/mpid/ch4/shm/posix/posix_coll.h

```diff
@@ -13,7 +13,8 @@
 
 #include "posix_impl.h"
 #include "ch4_impl.h"
-#include "ch4_coll_select.h"
+#include "ch4r_proc.h"
+#include "ch4_coll_impl.h"
 #include "posix_coll_params.h"
 #include "posix_coll_release_gather.h"
```

Comment:
@pavanbalaji In this file `posix_coll.h`, `release_gather` case for Bcast, Reduce, and Allreduce should be protected, as it does not run when threading is enabled (isThreaded=1). Earlier in file `posix_coll_select.h` (which was deleted in PR #4319), I had this code, which is necessary-

```    if (MPIR_CVAR_ENUM_IS(BCAST_POSIX_INTRA_ALGORITHM, release_gather)) {	
        /* release_gather based algorithm can be used only if MPICH is	
         * not multi-threaded */	
#ifdef MPICH_IS_THREADED	
        if (!MPIR_ThreadInfo.isThreaded) {	
            /* MPICH configured with threading support but not actually used */	
            return &MPIDI_POSIX_Bcast_intra_release_gather_cnt;	
        }	
#else	
        /* MPICH not configured with threading support */	
        return &MPIDI_POSIX_Bcast_intra_release_gather_cnt;	
#endif /* MPICH_IS_THREADED */	
    }
```

Reply:
Also, Allreduce needs additional constraint checking on non-commutative ops. `release_gather` cannot be called otherwise. 

PR #4319 broke this https://jenkins-pmrs.cels.anl.gov/job/mpich-master-ch4-ofi/compiler=gnu,fabric_prov=sockets,jenkins_configure=shm-coll,label=centos32/

Reply:
@jain-surabhi-23 I think we accidentally removed those checks in [487b72543e70].  I'll add them back.  Thanks for catching it.

Reply:
Fixed.